### ISSUE TITLE
Cyclic inheritance build error

### ISSUE THREAD

**Issue by arpie42** [blitz-research/monkey2#131]
Is this fixable?

File 'cycletest.monkey2' :
```
#Import "external"

Class Base
End

Class Derived Extends External
End

Function Main()
End
```
File 'external.monkey2' :
```
Class External Extends Base
	Method New()
	End
End Class

```
Result :

```
Parsing...
Semanting...
Translating...
Compiling...
Build error: System command 'g++ -I"/home/arpie/Coding/Monkey/Monkey2/modules/" -I"/home/arpie/Coding/Monkey/Monkey2/modules/monkey/native" -I"/home/arpie/Projects/BiaB/MazeM2/" -std=c++11 -c -o "/home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/build/_1src_2cycletest_0external.cpp.o" "/home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/cycletest_external.cpp"' failed.

g++ -I"/home/arpie/Coding/Monkey/Monkey2/modules/" -I"/home/arpie/Coding/Monkey/Monkey2/modules/monkey/native" -I"/home/arpie/Projects/BiaB/MazeM2/" -std=c++11 -c -o "/home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/build/_1src_2cycletest_0external.cpp.o" "/home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/cycletest_external.cpp"

In file included from /home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/../include/cycletest_external.h:12:0,
                 from /home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/cycletest_external.cpp:2:
/home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/../include/cycletest_cycletest.h:40:35: error: invalid use of incomplete type ‘struct t_default_External’
 struct t_default_Derived : public t_default_External{
                                   ^~~~~~~~~~~~~~~~~~
In file included from /home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/cycletest_external.cpp:2:0:
/home/arpie/Projects/BiaB/MazeM2/cycletest.buildv1.1.02/linux_debug/src/../include/cycletest_external.h:7:8: note: forward declaration of ‘struct t_default_External’
 struct t_default_External;
        ^~~~~~~~~~~~~~~~~~


***** Fatal mx2cc error *****

Internal mx2cc build error

```

**Comment by arpie42** [comment]
The issue I reported is the simplest form of the error that I can find.  However, in order to demonstrate why I want the External class to be in a separate file, this is what I am actually trying to do :

File 'cycletest.monkey2' :

```
#Import "mixin"

Class Base
End

Class Derived Extends Mixin<Base>
End

Function Main()
End
```

File 'mixin.monkey2' :

```
Class Mixin<T> Extends T
	Method New()
	End
End Class
```

This results in the same build error as reported previously but this may be a slightly harder case to get working?

**Comment by XANOZOID** [comment]
## **TLDR;**
The most straightforward approach to fix, at least the first basic example, would be to create individual header files for each user defined object( cpp struct ). 

## My analysis 
What's happening is that two separate header files are generated and the compiler doesn't know how to inherit from a forward declaration without knowing the full description of the parent class's members. 

2: But another, not so great, fix seems to be keeping the classes defined in the same monkey2 file. _(come back to this in a second )_

The problem actually occurs from inside of the imported file. So because the class from `File B` inherits from the first class in `File A` it `imports` File A and successfully gets the first class. The problem is that because `File A` was included its entire file will be processed ( I think ) and because `File B` needs `File A` to be processed the second class which relies on the unfinished `File B` will be unsuccessful. 
- (referencing 2 ) Which is why I believe keeping it in the same monkey2 works but not separated. 

## What happens when you have `Derived` class in a different file from `Base`? + final thoughts
Like I would have assumed, it fixes the issue. At a high level this seems like a straightforward fix (but at a lower level this may cause an unearthly amount of problems!).  We can either hopefully keep all elements of a same namespace inside of one large file or, probably a better option, have separated files for user defined structures. 

I hope this helps fixing the problem -- maybe I could even take a shot at it :)
